lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+type tokenFactory struct {
+	token string
+	mutex *sync.RWMutex
+}
+
+// powerFlexClient holds the PowerFlex HTTP client and an access token factory.
+type powerFlexClient struct {
+	driver *powerflex
+	token  *tokenFactory
+}
+
+// powerFlexStoragePool represents a storage pool in PowerFlex.
+type powerFlexStoragePool struct {
+	ID                 string `json:"id"`
+	Name               string `json:"name"`
+	ProtectionDomainID string `json:"protectionDomainId"`
+}
+
+// powerFlexProtectionDomain represents a protection domain in PowerFlex.
+type powerFlexProtectionDomain struct {
+	ID       string `json:"id"`
+	Name     string `json:"name"`
+	SystemID string `json:"systemId"`
+}
+
+// powerFlexProtectionDomainSdsRelation represents an SDS related to a protection domain in PowerFlex.
+type powerFlexProtectionDomainSdsRelation struct {
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	IPList []struct {
+		IP string `json:"ip"`
+	} `json:"ipList"`
+}
+
+// powerFlexSdc represents a SDC in PowerFlex.
+type powerFlexSdc struct {
+	ID       string `json:"id"`
+	Name     string `json:"name"`
+	HostType string `json:"hostType"`
+	NQN      string `json:"nqn"`
+}
+
+// powerFlexVolume represents a volume in PowerFlex.
+type powerFlexVolume struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	VolumeType    string `json:"volumeType"`
+	MappedSdcInfo []struct {
+		SdcID    string `json:"sdcId"`
+		SdcName  string `json:"sdcName"`
+		NQN      string `json:"nqn"`
+		HostType string `json:"hostType"`
+	} `json:"mappedSdcInfo"`
+}
+
+// Error tries to return all kinds of errors from the PowerFlex API in a nicely formatted way.
+func (p *powerFlexError) Error() string {
+	var errorStrings []string
+	for k, v := range *p {
+		errorStrings = append(errorStrings, fmt.Sprintf("%s: %v", k, v))
+	}
+
+	return strings.Join(errorStrings, ", ")
+}
+
+// ErrorCode extracts the errorCode value from a PowerFlex response.
+func (p *powerFlexError) ErrorCode() float64 {
+	code, ok := (*p)["errorCode"].(float64)
+	if !ok {
+		return 0
+	}
+
+	return code
+}
+
+// HTTPStatusCode extracts the httpStatusCode value from a PowerFlex response.
+func (p *powerFlexError) HTTPStatusCode() float64 {
+	code, ok := (*p)["httpStatusCode"].(float64)
+	if !ok {
+		return 0
+	}
+
+	return code
+}
+
+// newTokenFactory creates a new instance of an access token factory.
+func newTokenFactory() *tokenFactory {
+	return &tokenFactory{
+		mutex: &sync.RWMutex{},
+	}
+}
+
+// Get returns the current access token of the factory.
+func (t *tokenFactory) Get() string {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+	return t.token
+}
+
+// Set creates a new access token in the factory.
+// An expiry can be set after which the access token will be removed from the factory.
+func (t *tokenFactory) Set(token string, expiry time.Duration) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	go func() {
```

Comment:
Do we need a go routine here or can we store the expiry time internally and check for expiration in the Get function instead?

Reply:
Like the idea, haven't thought about it. Much simpler this way.

Reply:
Implemented as suggested using a check when retrieving the token.